### ISSUE TITLE
Let's use memento pattern to reduce the brittleness of serialization

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#814]
In short, apply proxy-serialization pattern:
http://blog.codefx.org/design/patterns/serialization-proxy-pattern/
#614 for some background.

Tasks:
- Give a shot at memento pattern as described in: #614 (https://youtu.be/V1vQf4qyMXg?t=56m12s)
- After doing one, we'll see how we can apply the pattern to our code base, have some time to reflect on it, and then apply it more broadly.

Goals:
- When done, we will be able to change any private field without worrying about breaking serialization

**Comment by RoiEXLab** [comment]
@DanVanAtta I agree, that this is a very useful approach in order to achieve cross version compatibility, the question is, what effort is needed to implement this...
As of now, there are 62 classes implementing serializable, not counting their subclasses, with 4 classes overriding readResolve or writeReplace in total.
Fun Fact: [History.java](https://github.com/triplea-game/triplea/blob/master/src/games/strategy/engine/history/History.java) is already using a SerializationProxy...

**Comment by DanVanAtta** [comment]
We've a goal to enable the loading and playing of save games, even if they were saved with an old game engine. To do this, the current thoughts are to save in a format we control, YAML, XML or JSON perhaps. 

It's a big step to get from where we are now, to there. My thoughts are then we can use memento to control how we read/write fields. The first pass will simply allow us to modify variable names safely. A second pass would then return values formatted in our custom, text based format.

Overall, it's a big step, but an important one. If we can accomplish the goal of having save games be loaded in new engine versions, it would allow us to release often. We could also remove all of the coded logic for launching 'old' jars. Last, with explicit save/load code, the code would become a lot less brittle, we could change variable names safely.

**Comment by ssoloff** [comment]
@DanVanAtta @ron-murhammer @RoiEXLab 

I'm almost done with my initial prototype that addresses this issue (I hope to have it ready before the end of the weekend).  While the framework for implementing the Serialization Proxy pattern is fairly complete, I haven't gone too far down the rabbit hole of adding proxies for every `Serializable` class that will require it because I would like your feedback to make sure I'm heading in the right direction.

Where is the best place to start this discussion?  Should I just do it in this issue and link to my spike branch?  Or would it be easier for me to open a PR that would be for discussion only and never merged?

The initial set of changes I am proposing are fairly large: there's about 2,000 new lines of non-test code.  However, I don't think a line-by-line review is required at this time, as I'm only looking for high-level architectural feedback.

**Comment by ssoloff** [comment]
@DanVanAtta @ron-murhammer @RoiEXLab

(I'm adding the initial review of my prototype here.  Please advise if I should open an ephemeral PR instead.)

### Introduction

I've created a [branch](https://github.com/ssoloff/triplea/tree/spike-issue-814-introduce-serialization-proxies) to prototype the addition of the Serialization Proxy pattern.  I thought this problem was complicated enough without adding the additional concern of changing persistence formats.  Therefore, I've focused on the first part of the problem, which is namely splitting the persistence layer concerns from the domain model, while continuing to use Java serialization as the persistence format.  @DanVanAtta seemed to suggest a similar two-step approach in [this comment](https://github.com/triplea-game/triplea/issues/1036#issuecomment-238965162).

The prototype is not complete in that one cannot persist a complete `GameData` instance with it at the moment.  However, the heavy lifting to introduce a framework for the Serialization Proxy pattern is complete.  I have only implemented a few serialization proxies as a proof of concept.  The purpose of this review is to give the dev team a high-level overview of how it works.  **I would like to know if I'm going down the wrong path** so I can readjust before plowing ahead with writing serialization proxies for every type required to persist `GameData.`

### Architectural overview

#### Serialization proxy framework

I introduced a framework for realizing the Serialization Proxy pattern in the `g.s.persistence.serializable` package.  An UML class diagram for this framework is given below:

![serialization-proxy-class](https://cloud.githubusercontent.com/assets/4826349/25369002/a78f63be-294e-11e7-9030-6cab4f5293a2.png)

`Foo` represents a non-serializable domain model class with public properties `bar` and `baz` that should be persisted.  In order to serialize such a class, an appropriate `PersistenceDelegate` must be defined that knows how to marshall and unmarshall `Foo` instances to and from a serializable representation.  In this case, the serializable representation is the `FooProxy` class.

The `ObjectInputStream` and `ObjectOutputStream` classes are specializations of the corresponding classes in `java.io` that know how to delegate serialization responsibilities for non-serializable classes to `PersistenceDelegate`s.  If a `PersistenceDelegate` is registered for a class, the streams will delegate the core operations `annotateClass()`, `replaceObject()`, `resolveClass()`, and `resolveObject()` to the `PersistenceDelegate`.  Otherwise, they simply delegate to the default behavior in `java.io`, such that already-serializable classes are handled without change.

You can see an example of how everything works together in [`ObjectInputOutputStreamIntegrationTest`](https://github.com/ssoloff/triplea/blob/spike-issue-814-introduce-serialization-proxies/src/test/java/games/strategy/persistence/serializable/ObjectInputOutputStreamIntegrationTest.java).  In this test, `Foo` is represented by `FakeNonSerializableClass`.

##### Versioning

The above framework successfully decouples the persistence layer from the domain model.  Thus, `Foo` can be refactored without having to worry about breaking serialization.  However, if a change to `Foo` requires a corresponding change in the persistence layer (e.g. changing the type of an existing property or adding a new property that must be persisted), the `FooProxy` class must be changed.  Thus, the same issues that used to plague `Foo` now affect `FooProxy`, and there needs to be some way for serialization proxies to support versioning when necessary.

I chose not to impose many constraints on serialization proxies, including requiring them to handle versioning beyond what the Java serialization framework already implements (i.e. `serialVersionUID`, type-compatible fields, etc.).  If that's good enough, then they can use it.  But I suspect there will be many TripleA proxies that require something more robust.  For those cases, I introduced the helper class [`VersionedProxySupport`](https://github.com/ssoloff/triplea/blob/spike-issue-814-introduce-serialization-proxies/src/main/java/games/strategy/persistence/serializable/VersionedProxySupport.java).

`VersionedProxySupport` requires proxies to implement `Externalizable` instead of `Serializable` to bypass certain constraints in the Java serialization framework.  It uses reflection to find the appropriate methods in the proxy class to read and write each supported version.  (The choice to use reflection to accomplish this was arbitrary.  In fact, I chose a totally different solution when implementing versioning for mementos [see below].  We can compare the two approaches and use whichever one the team is more comfortable with.)  You can see an example of how a versioned serialization proxy would handle multiple versions in  [`VersionedProxyTest`](https://github.com/ssoloff/triplea/blob/spike-issue-814-introduce-serialization-proxies/src/test/java/games/strategy/persistence/serializable/VersionedProxyTest.java).

##### Application to TripleA

Ideally, if TripleA ran in an IoC container or something with at least a service bus, there would be a `PersistenceDelegateRegistry` service.  The Whiteboard pattern could then be used to register `PersistenceDelegate`s from various disparate modules.  As that's not the case, I'm currently using a [factory](https://github.com/ssoloff/triplea/blob/spike-issue-814-introduce-serialization-proxies/src/main/java/games/strategy/engine/persistence/serializable/PersistenceDelegateRegistryFactory.java) that knows how to create a `PersistenceDelegateRegistry` with the required registrations that is appropriate for serializing instances of `GameData`.  It's a bit of a monolithic smell for me, but I don't think it can be helped as TripleA is a monolithic application.

You can see a real example of a serialization proxy that supports versioning (albeit with only one version) for the `g.s.util.Version` class in [`VersionPersistenceDelegate`](https://github.com/ssoloff/triplea/blob/spike-issue-814-introduce-serialization-proxies/src/main/java/games/strategy/internal/persistence/serializable/VersionPersistenceDelegate.java) and [`VersionProxy`](https://github.com/ssoloff/triplea/blob/spike-issue-814-introduce-serialization-proxies/src/main/java/games/strategy/internal/persistence/serializable/VersionProxy.java).

#### Memento framework

Even though it was discussed in #814 and #1036, I originally did not think it would be necessary to introduce the Memento pattern.  Once I had the serialization proxies in place, I didn't see a need to introduce an intermediate representation between the domain object and the proxy.  Unfortunately, my reasoning assumed there was a single representation for each domain object in the persistence layer.  This assumption turned out to be false when I started looking more closely at how different areas of the code use `GameData`.

There is the straightforward, and what I'll call "complete," representation used for saving games.  However, there is a second representation that includes everything except delegates.  This is apparently used when certain areas of the code (e.g. some odds calculators) clone `GameData` for their own purposes.  As serialization proxy selection is based on an object's class, it wasn't possible to choose between the representation with or without delegates when writing an object via `ObjectOutputStream` without adding some bad smelling properties to `GameData` (e.g. `boolean shouldPersistDelegates()`).

So taking a step back, it seemed like using a Memento for `GameData`, at least in this situation, made sense.  For no particular reason, I didn't want to add a `getMemento()` or similar method(s) to `GameData` in order to keep the memento/originator conversion logic outside of the domain object.  Instead, I introduced a general framework for mementos in the `g.s.util.memento` package.  An UML class diagram for this framework is given below:

![memento-class](https://cloud.githubusercontent.com/assets/4826349/25368882/e689c75e-294d-11e7-9b87-03d4e1fdfff7.png)

Just as with serialization proxies, the idea here is to keep the memento/originator conversion (import/export) out-of-band with the associated domain object (the originator).  The importer/exporter should rely on the public (or at least non-private) API of the domain object.

Mementos themselves are just opaque representations of state and, in and of themselves, are not generally persistable.  Therefore, in order to serialize and deserialize them, each concrete memento implementation will require a serialization proxy, as discussed in the previous section.  To make this a bit clearer, I added a parallel [serializable memento importer/exporter framework](https://github.com/ssoloff/triplea/tree/spike-issue-814-introduce-serialization-proxies/src/main/java/games/strategy/persistence/memento/serializable) that merges the concepts of the memento and serialization proxy frameworks into a single concern.

I added one concrete memento implementation: [`PropertyBagMemento`](https://github.com/ssoloff/triplea/blob/spike-issue-814-introduce-serialization-proxies/src/main/java/games/strategy/util/memento/PropertyBagMemento.java).  This is just a generic property bag (it maps `String` keys to `Object` values).

##### Versioning

Just as with the serialization proxies, mementos may need to be versioned.  Again, I didn't want to put this constraint on the general memento framework, so I leave it up to different memento implementations to implement it as necessary.  `PropertyBagMemento` supports the notion of a "schema" (i.e. what each property "means," what properties must exist, etc.), and a schema includes a monotonically-increasing version to support forward migration.  The versioning support is provided through the [`PropertyBagMementoImporter`](https://github.com/ssoloff/triplea/blob/spike-issue-814-introduce-serialization-proxies/src/main/java/games/strategy/util/memento/PropertyBagMementoImporter.java) and [`PropertyBagMementoExporter`](https://github.com/ssoloff/triplea/blob/spike-issue-814-introduce-serialization-proxies/src/main/java/games/strategy/util/memento/PropertyBagMementoExporter.java) classes.

Unlike the approach I took with serialization proxies where I used reflection to discover the methods used to read and write each version, for `PropertyBagMemento` I am explicitly providing what basically amounts to a map of handlers for importing and exporting different versions.  Again, I did this to allow the team to compare the two approaches.  Other than how the handlers are discovered, the versioning mechanism is very similar to what I did for `VersionedProxySupport` in the serialization proxy framework.

##### Application to TripleA

As I mentioned earlier, the need for the Memento pattern was due to how different areas of the code save and/or clone `GameData`.  To illustrate this, I implemented a [factory](https://github.com/ssoloff/triplea/blob/spike-issue-814-introduce-serialization-proxies/src/main/java/games/strategy/engine/data/GameDataMemento.java) that can (eventually) create various types of mementos for a `GameData` instance.  Note that there is no serialization proxy for `GameData` itself because a `GameData` instance should not be persisted--only its memento should be persisted.

To bring all this stuff (serialization proxy and memento frameworks) together, I added two new methods to `GameDataManager` for [loading](https://github.com/ssoloff/triplea/blob/spike-issue-814-introduce-serialization-proxies/src/main/java/games/strategy/engine/framework/GameDataManager.java#L286) and [saving](https://github.com/ssoloff/triplea/blob/spike-issue-814-introduce-serialization-proxies/src/main/java/games/strategy/engine/framework/GameDataManager.java#L320) `GameData` in serializable format.  That should give you a good idea of how this all works from the top.

### Next steps

What I'm primarily looking for from this review is to ensure I'm heading down the right architectural path.

Secondarily, I'd appreciate any general feedback, such as improving package, type, and member names.

**Comment by ssoloff** [comment]
@DanVanAtta @ron-murhammer @RoiEXLab

Bumping previous comment...

**Comment by RoiEXLab** [comment]
> Secondarily, I'd appreciate any general feedback, such as improving package, type, and member names.

@ssoloff I think `games.strategy.*` should be changed to `org.triplea` as discussed in #1598, I'm fine with the other names...

Your plans look good, although I still think we shouldn't use serialization at all and just encode objects into bytestreams manually...

**Comment by ssoloff** [comment]
Thanks for the feedback, @RoiEXLab.

> I think games.strategy.* should be changed to org.triplea as discussed in #1598, I'm fine with the other names...

I'm good with that.  I was following the discussion in #1598 and actually started to name my packages `org.triplea.*` in this spike, but I wasn't sure if it would be too disruptive, so I backed out.  If @DanVanAtta and @ron-murhammer don't state any objection, I'll go ahead and rename the packages.

> I still think we shouldn't use serialization at all and just encode objects into bytestreams manually...

Yeah, this is tough.  Like I said above, I didn't want to change two things at once (but I will if that's what the team wants).  If it helps any, using `Externalizable` is essentially "encoding objects into byte streams manually."  The only thing not under our control is encoding/decoding the object type, and thus we also lose control of object instantiation during decoding.  That is, the proxies are required to have a public no-arg constructor so they can be instantiated via reflection.  But the proxies are not intended for normal use (hence why they are in an `internal` package), and no constraints are placed on the domain classes.

**Comment by ssoloff** [comment]
@DanVanAtta @ron-murhammer Bumping...

If possible, I'd like to implement this while I still have some contiguous free time available over the next few weeks.  I'd appreciate any feedback either of you can provide on [this comment](https://github.com/triplea-game/triplea/issues/814#issuecomment-296905464) before I begin in earnest.

**Comment by DanVanAtta** [comment]
One thing to call out, we're committing to a pretty large effort that also requires a revision bump. Before we do that, we need to get everyone on an upgrade track. The earlier we start the better. My understanding is until we get users on a patched version, the common version being played now freezes when we bump the game version number.  That may be a potential problem for anything that requires a version number updates.

Re: package rename

I'm in favor of the rename since this change would need a version bump anyways. Only concern I would have are merge conflicts. Renaming the packages as late as possible but before final testing may potentially minimize those.


Some quick responses for now, I'll try to dig into this in more detail soon (next day or two) and get back to you.

**Comment by ssoloff** [comment]
Thanks, @DanVanAtta.

> My understanding is until we get users on a patched version, the common version being played now freezes when we bump the game version number.

Yes, #1558 has not yet been released.

> I'm in favor of the rename since this change would need a version bump anyways.

Are you envisioning an across-the-board package rename with this change?  I'll admit that when @RoiEXLab made his suggestion, that's not what I imagined.  I thought I would simply add the **new** types to new packages following the new naming convention but **old** types would be left where they are for now.

Once the serialization proxies are in place, we would be free to rename the packages of all existing types without consequence (i.e. no additional version bump required).  Therefore, my preference would be to do the mass package rename of existing types in a separate PR after this feature has been fully merged, tested, and released.

**Comment by ron-murhammer** [comment]
@ssoloff Looks pretty good to me. I agree with keeping the package renaming separate from this change and do that as essentially the last PR before we look to do final testing and release. But yes the scope of that would be across-the-board moving most packages to start with `org.triplea` or `org.tripleagame` as discussed in #1598.

If you have specific questions or points you'd like to discuss please highlight them. Nice work so far.

**Comment by DanVanAtta** [comment]
:cry:  I can't believe these two PRs were not linked to this issue or mentioned sooner:

https://github.com/triplea-game/triplea/pull/945
https://github.com/triplea-game/triplea/pull/947

Noting: https://github.com/triplea-game/triplea/pull/947/files#diff-b288011734ac39f293bc8d127fe98662R50

Which has already been deployed on the last major release.

**Comment by ssoloff** [comment]
Thanks for the references, @DanVanAtta.

I agree that the approach taken in the PRs you cited satisfies the most important goals of employing the serialization proxy pattern.  Specifically, the ability to refactor domain class internals without breaking serialization.

However, there are a few areas I'm concerned about:

<hr>

#### Feature toggle to support beta testing

Because this feature is so large, I'm assuming I **will** get something wrong because there is just so many permutations to test.  Therefore, I don't want to roll it out in one "big bang" PR whereby it replaces the current serialization mechanism and then have to deal with a bunch of fire fighting that may potentially block devs from doing other work and users from using pre-release versions.

I envisioned adding a feature toggle that would give users the option of using the new framework for loading/saving games and serializing objects over the wire.  By default, the feature toggle would be disabled and the existing serialization code would be used.  Only if someone wanted to explicitly beta test the new serialization framework would it be enabled.  Once we are sure the new framework is working (and we are ready to bump versions), the feature toggle and the old serialization code can be removed.  (Yes, that means that during the development of this feature, the existing `Serializable` domain classes would not be modified, except maybe to add new APIs to support creation/access methods required by the external persistence framework.)

With the approach described in the above PRs, this would require checking the feature toggle in each `writeReplace()` method.  That is potentially tens (or hundreds?; I haven't yet counted the number of classes that are serialized) of locations where the feature can be toggled.  Whereas with the proposed solution, the feature toggle only needs to be checked where `ObjectOutputStream` and `ObjectInputStream` are used (ideally this is four locations: two for saving/loading games and two for sending/receiving objects via RMI).

#### Refactoring

In the above PRs, the serialization proxy is nested within the domain class.  This makes it difficult (but not impossible) to rename the domain class and/or move it to a different package.  The fully-qualified serialization proxy class name needs to remain unchanged to, for example, avoid breaking old save games.  One can get around this by refactoring the domain class as desired, but leaving its "shell" behind in the original location to act as a trampoline for the old serialization proxy.  That is, the original domain class would be empty except for the nested serialization proxy, whose `readResolve()` method would redirect to the new domain class (which presumably now has a "new" serialization proxy and a `writeReplace()` method that uses the new proxy).  Introducing such concerns into the domain class just feels weird.

With the proposed solution, performing such refactorings on the domain class are straightforward.  Obviously, the serialization proxy itself suffers the same limitations described in the previous paragraph, but it is less likely that these standalone serialization proxies will be candidates for such refactoring.

#### Persistence ignorance

This is nitpicking, but in the above PRs, the domain class is not truly "persistence ignorant".  I mentioned this in the Refactoring topic above: the domain class still contains intimate details about a specific persistence framework (capital-S Serialization, in this case).  If one wants to add support for an alternate persistence framework in the future that could be used in parallel with the existing framework, the domain class would now contain knowledge of multiple persistence frameworks.  Anything related to persistence seems like a responsibility that should be external to the domain class.

<hr>

In summary, I have no problem implementing this feature based on the work that has already been done using `@SerializationProxySupport`.  As I said, that solves what is probably my biggest pet peeve with the codebase right now: the inability to freely refactor classes without having to worry about breaking backwards compatibility.

However, I would ask that you consider the issues I raised above.  It's possible they are not important considerations for TripleA and that I simply over-engineered the solution.  I'm ok with whatever approach you want to take.  I just want to get to work on this feature. :smiley:

**Comment by ssoloff** [comment]
@DanVanAtta Any final thoughts on what path you would like me to take based on my [previous comment](https://github.com/triplea-game/triplea/issues/814#issuecomment-302999809)?  I'm probably going to start this weekend and would like to avoid back-tracking if possible.

**Comment by DanVanAtta** [comment]
Sry for late response.

I see we are moving the infra to do the proxy to a new class. Sounds okay. I would want to verify that the link between proxy and proxied class is made clear, preferably compile time checked.

One question, how will the versioning be useful? I was thinking we would stick to only using major version to check compat. Just want to be sure that any extra overhead here is worth it.

Re: gatekeeping and going wholesale

This might be a place where it makes sense to go whole sale. If we do a feature flag, then we'll have a check the world scenario when it is turned on. On the other hand if we do 3-5 classes at a time, the testing scope will remain limited. Doing the proxy for the other classes, it was somewhat obvious when it was broken. Easiest though IMO if we can focus on a few classes and not whole game data.

Also to that end, a very notable benefit I found when doing proxy was that we could drop the GameData parameter. This made things a good deal nicer, I look forward to when we can create simple Triplea unit objects without needing a game data.  I was wondering if you were considering this? Is this reason to do incremental conversions, or would we maybe do a second pass to remove the game data when we have everything proxied?

**Comment by ssoloff** [comment]
@DanVanAtta 

> I would want to verify that the link between proxy and proxied class is made clear, preferably compile time checked.

Yes, there is a compile-time reference from the proxy class to the domain class, but not vice versa (as it should be, IMO).

> how will the versioning be useful?

First, let me clarify that this design was proposed before #1739 was opened, so the design may need to be changed.  Second, the proposed versioning solution is totally optional--there is no requirement that it be used by any of the proxies.

To address the core of your question, I originally included the ability to version the proxies because, in all the various issues that were opened to discuss serialization proxies, new network protocols, etc., I latched on to a common theme that I'll paraphrase as, "we need a format we can control so that games from old engines can be loaded."  The definition of "old" is obviously the issue here.  I guess I took it to mean, "however old you want it to be," versus what #1739 seems to be going for: "since the last major release."

I guess what proxy versioning buys you is that you can punt on bumping the product's major version in light of what would normally be viewed as an incompatible change.  If the proxy can handle what would otherwise be an incompatible change, it delays making a new major product release.  I honestly have no idea if that should even be a concern or how frequent major releases will become in the future.

One more thing...  A versioned proxy (in the context of the implementation I've proposed) that only supports a single version is essentially the same as one that doesn't support versioning at all.  So, if we decide that versioning proxies is not necessary, it will be pretty easy to rip out (basically changing the signature of the method that is equivalent to `readObject()`; all the code remains the same).

> This might be a place where it makes sense to go whole sale. 

Let me state up front that I agree with you and definitely want to do this across multiple PRs to avoid a huge review.  Submitting a PR with 3-5 classes at a time, as you suggested, is pretty much the range I had in mind.

Heh, I actually thought by using a feature toggle I was addressing what is essentially your concern.  Obviously, the feature toggle approach requires that people actually turn it on to test incrementally.  Otherwise, as you said, there will be a big bang test at the end, which I agree is not desirable.

**Unless I'm misunderstanding something, the issue I have with not using the feature toggle is that every time I submit a PR, save games prior to the merge will be broken.  In addition, until the entire feature is in place, it will delay the next official product release.  My current estimate is that this will take me about 2-4 weeks.  And if something comes up and I have to step away for a bit half way though the feature, I don't want to be responsible for holding up the release. :smile:**

Also, if we go down the route of breaking compatibility, then I'll have to wait until the next release to submit the first PR, correct?  My understanding is we want the next release to be backwards compatible with 1.9.0.0.3635 save games. 

> a very notable benefit I found when doing proxy was that we could drop the GameData parameter.

I haven't yet gone further than the few proxies I mentioned in my original overview several weeks ago, so I haven't seen this side effect yet.  I will keep it in mind and definitely work towards that goal if it is possible.

**Comment by DanVanAtta** [comment]
Thanks for the responses and picking this up @ssoloff . Closing this issue in favor of #1889 and other work. We may want to create a github to track this. As is, I think this tracking issue has ran its course. Closing for now.
